flink/src/main/java/org/apache/iceberg/flink/sink/FlinkAppenderFactory.java

```diff
@@ -0,0 +1,231 @@
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public DataWriter<RowData> newDataWriter(EncryptedOutputFile file, FileFormat format, StructLike partition) {
+    return new DataWriter<>(
+        newAppender(file.encryptingOutputFile(), format), format,
+        file.encryptingOutputFile().location(), spec, partition, file.keyMetadata());
+  }
+
+  @Override
+  public EqualityDeleteWriter<RowData> newEqDeleteWriter(EncryptedOutputFile outputFile, FileFormat format,
+                                                         StructLike partition) {
+    MetricsConfig metricsConfig = MetricsConfig.fromProperties(props);
+    try {
+      switch (format) {
+        case AVRO:
+          return Avro.writeDeletes(outputFile.encryptingOutputFile())
+              .createWriterFunc(ignore -> new FlinkAvroWriter(lazyEqDeleteFlinkSchema()))
+              .withPartition(partition)
+              .overwrite()
+              .setAll(props)
+              .rowSchema(eqDeleteRowSchema)
+              .withSpec(spec)
+              .withKeyMetadata(outputFile.keyMetadata())
+              .equalityFieldIds(equalityFieldIds)
+              .buildEqualityWriter();
+
+        case PARQUET:
+          return Parquet.writeDeletes(outputFile.encryptingOutputFile())
+              .createWriterFunc(msgType -> FlinkParquetWriters.buildWriter(lazyEqDeleteFlinkSchema(), msgType))
+              .withPartition(partition)
+              .overwrite()
+              .setAll(props)
+              .metricsConfig(metricsConfig)
+              .rowSchema(eqDeleteRowSchema)
+              .withSpec(spec)
+              .withKeyMetadata(outputFile.keyMetadata())
+              .equalityFieldIds(equalityFieldIds)
+              .buildEqualityWriter();
+
+        default:
+          throw new UnsupportedOperationException("Cannot write unknown file format: " + format);
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public PositionDeleteWriter<RowData> newPosDeleteWriter(EncryptedOutputFile outputFile, FileFormat format,
+                                                          StructLike partition) {
+    MetricsConfig metricsConfig = MetricsConfig.fromProperties(props);
+    try {
+      switch (format) {
+        case AVRO:
+          return Avro.writeDeletes(outputFile.encryptingOutputFile())
+              .createWriterFunc(ignore -> new FlinkAvroWriter(lazyPosDeleteFlinkSchema()))
+              .withPartition(partition)
+              .overwrite()
+              .setAll(props)
+              .rowSchema(posDeleteRowSchema)
+              .withSpec(spec)
+              .withKeyMetadata(outputFile.keyMetadata())
+              .buildPositionWriter();
+
+        case PARQUET:
+          RowType flinkPosDeleteSchema = FlinkSchemaUtil.convert(DeleteSchemaUtil.posDeleteSchema(posDeleteRowSchema));
```

Comment:
@rdblue   Here I use the parquet writer which is built from <path, pos, row> schema, rather than the writer built from `row` schema.   

I tried to hide all the <path, pos, row> write logic inside the `Parquet#buildPositionWriter`, but failed to make it work.  I passed a function to this builder by creating a  parquet writer built by `rowSchema`, and construct a PositionDeleteStructWriter with a path string writer,  a pos long writer and the  row struct writer,   it does not work because the `currentPath` for [here](https://github.com/apache/iceberg/blob/f950a3e63c98e88b1f8905faacf33a1d5d31e4c0/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetWriters.java#L126) has been messed up.   Still figure out how to make this work.

Reply:
Let me provide more details, assume that we want to build a pos delete writer with row schema <id long, data string>.

The straightforward way is to build a struct writer like this:

```java
    PositionDeleteStructWriter
               |
               |---------->  StringWriter (file_path field)
               |---------->  LongWriter   (pos       field)
               |
               |
               |---------->  StructWriter (row  field)
                                 |----------> LongWriter    (id   field)
                                 |----------> StringWriter  (data field)

```

The question is how to build the PositionDeleteStructWriter, if we just provide a function to create the `row` field's `StructWriter` for flink, and then concat the <StringWriter, LongWriter, StructWriter> to PositionDeleteStructWriter constructor,
then the problem is:  the leaf node `ColumnWriter` will have an incorrect `ColumnDescriptor`:

```java
     PositionDeleteStructWriter
               |
               |---------->  StringWriter {<file_path>,type,maxRep,maxDef}
               |---------->  LongWriter   {<pos>,type,maxRep,maxDef}
               |
               |
               |---------->  StructWriter 
                                 |----------> LongWriter    {<id>, type, maxRep, maxDef}
                                 |----------> StringWriter  {<data>,type, maxRep, maxDef}
```

In fact, for this example, the correct ColumnWriter should be:

```java
     PositionDeleteStructWriter
               |
               |---------->  StringWriter {<file_path>,type,maxRep,maxDef}
               |---------->  LongWriter   {<pos>,type,maxRep,maxDef}
               |
               |
               |---------->  StructWriter 
                                 |----------> LongWriter    {<row,id>, type, maxRep, maxDef}
                                 |----------> StringWriter  {<row,data>,type, maxRep, maxDef}
```

Reply:
In my mind, there are two ways to solve the problem, the first one is the current version: Just use the completed scheme which has included the `file_path` field and `pos` field to build the flink/spark struct writer, tradeoff
is exposing few details to upper, such as we need to convert the `CharSequence` to flink's `StringData` before we write the `file_path` value (see the `ParquetValueWriters.PathPosAccessor`).  The current version is easy to implement.

Another way is: call the `row` schema's `createWriterFunc` in parquet visitor when building the whole struct writer by traversing the columns. In that visitor, we could add the `row` prefix to the fieldNames and finally the ColumnDescriptor won't be messed up.
That seems to be more complex when compared to the first version. The good thing is hiding all the constructing details inside the builder.

IMO, I'd prefer to use the current version because the `FlinkAppenderFactory` and `SparkAppenderFactory` has already hidden the details for upper layer. People won't likely to write their own `AppenderFactory`, even if they want to write their own it will be easy when following the current flink/spark appender factory.

Reply:
Sorry, I also hit the column descriptor problem when I tried to update this. I have it working for Avro, but not for Parquet. I should have commented with the results of my work so you didn't try it as well.

I agree with you that this would be really difficult to solve and it is probably not worth it. We'll just have to go with the existing solution for Parquet until we get a better Parquet API that can load columns by ID rather than by path.

Reply:
> we get a better Parquet API that can load columns by ID rather than by path.

Sounds like it's the correct direction to solve this issue, will that need much work to accomplish (I'm not quite sure). 

BTW,  do you have other concern about this patch ?

Reply:
I'm reviewing it now. I just wanted to comment yesterday to let you know about the Parquet problem.

For the amount of work in Parquet, it shouldn't be too difficult for just this, but we need to support field IDs more generally over there and that is going to be a big project with all of the little things that need to be done.